src/Umbraco.Examine/ContentIndexPopulator.cs

```diff
@@ -70,31 +77,89 @@ protected override void PopulateIndexes(IReadOnlyList<IIndex> indexes)
+        {
             IContent[] content;
 
             do
             {
-                if (!_publishedValuesOnly)
-                {
-                    content = _contentService.GetPagedDescendants(contentParentId, pageIndex, pageSize, out _).ToArray();
-                }
-                else
+                content = _contentService.GetPagedDescendants(contentParentId, pageIndex, pageSize, out _).ToArray();
+
+                if (content.Length > 0)
                 {
-                    //add the published filter
-                    //note: We will filter for published variants in the validator
-                    content = _contentService.GetPagedDescendants(contentParentId, pageIndex, pageSize, out _,
-                        _publishedQuery, Ordering.By("Path", Direction.Ascending)).ToArray();
+                    // ReSharper disable once PossibleMultipleEnumeration
+                    foreach (var index in indexes)
+                        index.IndexItems(_contentValueSetBuilder.GetValueSets(content));
                 }
 
+                pageIndex++;
+            } while (content.Length == pageSize);
+        }
+
+        protected void IndexPublishedContent(int contentParentId, int pageIndex, int pageSize,
+            IReadOnlyList<IIndex> indexes)
+        {
+            IContent[] content;
+
+            var publishedPages = new HashSet<int>();
+
+            do
+            {
+                //add the published filter
+                //note: We will filter for published variants in the validator
+                content = _contentService.GetPagedDescendants(contentParentId, pageIndex, pageSize, out _, _publishedQuery,
+                    Ordering.By("Level", Direction.Ascending)).ToArray();
+               
                 if (content.Length > 0)
                 {
+                    var indexableContent = new List<IContent>();
+
+                    // get the max level in this result set
+                    int maxLevel = content.Max(x => x.Level);
+
+                    // gets the first level pages, these are always published because _publishedQuery filter
+                    // and store the id in a hash set so we can track them
+                    var firstLevelContent = content.Where(x => x.Level == 1);
+                    
+                    foreach (var item in firstLevelContent)
+                    {
+                        publishedPages.Add(item.Id);
+                        indexableContent.Add(item);
+                    }
+
+                    // get content per level so we can filter the pages that don't have a published parent
+                    // this because the paged descendants method does not take unpublished ancestors into account
+                    for (var level = 2; level <= maxLevel; level++)
+                    {
+                        var levelContent = content.Where(x => x.Level == level && publishedPages.Contains(x.ParentId));
```

Comment:
We cannot do this, this means we are iterating the entire result set again. We need to be able to do all of this without re-iterating the result set more than once (i.e. we only do a foreach on the result and we track what is needed within the single loop)

Reply:
Refactored to everything in a single foreach